### DISCUSSION TITLE
Weird emoji support status

### DISCUSSION THREAD

**Discussion by dcpurton** [neomutt/neomutt#3276] [Development]
This is just a discussion thread to track the current status of support for weird emojis.

The issues are not so much with Neomutt, but with `ncursesw` and `libc`.

**Comment by dcpurton** [comment]
A little test program to see where things are at.

Compile with `gcc test.c -lncursesw`.

```C
#define _XOPEN_SOURCE_EXTENDED
#define _XOPEN_SOURCE
#include <locale.h>
#include <ncurses.h>
#include <string.h>
#include <wchar.h>

// adapted from lynx source
#define IsSpecialAttrChar(a)	(((a) > '\002') && ((a) <= '\010'))
#define IsNormalChar(a) ((a) != '\0' && !IsSpecialAttrChar(a))
#define UCH(ch) ((unsigned char)(ch))
int getwidth(const char *string)
{
  int used = strlen(string);
  int result = 0;
  int maxCells = 50;
  WINDOW *fake_win = NULL;
  int fake_max = 102;

  fake_win = newwin(2, fake_max, 0, 0);
  if (fake_win != 0) {
    int new_x = 0;
    int new_y = 0;
    int x = 0;
    int n;

    wmove(fake_win, 0, 0);
    for (n = 0; n < used; ++n) {
      if (IsNormalChar(string[n])) {
        waddch(fake_win, UCH(string[n]));
        getyx(fake_win, new_y, new_x);
        if (new_y > 0 || new_x > maxCells)
          break;
        x = new_x;
      }
    }
    result = x;
    delwin(fake_win);
  }
  return result;
}

int main()
{
  setlocale(LC_ALL, "en_AU.UTF-8");
  initscr();
  printw("(🧑‍🔬) correct width: 4\n");
  printw("ncurses calculated width: %d\n", getwidth("(🧑‍🔬)"));
  printw("wcswidth calculated width: %d\n", wcswidth(L"(🧑‍🔬)", 50));
  printw("\n(👍🏻) correct width: 4\n");
  printw("ncurses calculated width: %d\n", getwidth("(👍🏻)"));
  printw("wcswidth calculated width: %d\n", wcswidth(L"(👍🏻)", 50));
  printw("\n(✒️) correct width: 4\n");
  printw("ncurses calculated width: %d\n", getwidth("(✒️)"));
  printw("wcswidth calculated width: %d\n", wcswidth(L"(✒️)", 50));
  printw("\n(🔓) correct width: 4\n");
  printw("ncurses calculated width: %d\n", getwidth("(🔓)"));
  printw("wcswidth calculated width: %d\n", wcswidth(L"(🔓)", 50));
  printw("\n(é) correct width: 3\n");
  printw("ncurses calculated width: %d\n", getwidth("(é)"));
  printw("wcswidth calculated width: %d\n", wcswidth(L"(é)", 50));
  refresh();
  getch();
  endwin();

  printf("printf output: (🧑‍🔬)(👍🏻)(✒️)(🔓)(é)\n");

  return 0;
}
```

**Comment by flatcap** [comment]
**System**:
- Fedora 35
- glibc 2.34
- ncurses 6.2-8.20210508

**Terminals**:
- alacritty 0.9.0
- kitty 0.24.2
- gnome-terminal 3.42.2

| Alacritty                                            | Gnome-Terminal                                   | Kitty                                            |
| :--------------------------------------------------- | :----------------------------------------------- | :----------------------------------------------- |
| ![a1](https://flatcap.org/mutt/emoji/alacritty1.png) | ![g1](https://flatcap.org/mutt/emoji/gnome1.png) | ![k1](https://flatcap.org/mutt/emoji/kitty1.png) |
| ![a2](https://flatcap.org/mutt/emoji/alacritty2.png) | ![g2](https://flatcap.org/mutt/emoji/gnome2.png) | ![k2](https://flatcap.org/mutt/emoji/kitty2.png) |

**Comment by dcpurton** [comment]
I notice [kitty](https://github.com/kovidgoyal/kitty) uses its own `wcswidth` implementation. This does better, but still fails with the ZWJ case (see issue https://github.com/kovidgoyal/kitty/issues/3810). Also of interest is https://github.com/kovidgoyal/kitty/issues/3998.

ncurses also uses its own implementation of `wcswidth`, but it doesn't seem to do anything special for emojis.

Some South Asian scripts also use ZWJ characters (https://en.wikipedia.org/wiki/Zero-width_joiner) and these are similarly poorly supported. I doubt anyone uses these complex layout scripts in any sort of terminal application though.

## Conclusion:

It's all a bad idea.
